core/src/main/java/org/apache/calcite/rel/RelCollationTraitDef.java

```diff
@@ -42,6 +48,54 @@ public class RelCollationTraitDef extends RelTraitDef<RelCollation> {
   public static final RelCollationTraitDef INSTANCE =
       new RelCollationTraitDef();
 
+  /** ThreadLocal to store the current planner during optimization.
+   * This allows trait satisfaction checking to access the planner's collation map
+   * without modifying the trait interface. */
+  private static final ThreadLocal<RelOptPlanner> CURRENT_PLANNER =
+      new ThreadLocal<>();
+
+  /** Maps planners to their original collation maps using weak references
+   * for planner keys to avoid memory leaks when planners are garbage collected.
+   * Each planner gets its own map to store optimization history for that query.
+   * The inner map uses HashMap because RelCollations are typically kept alive
+   * by the planner's RelSet and won't be GC'd during optimization. */
+  private static final Map<RelOptPlanner, Map<RelCollation, List<RelFieldCollation>>>
+      PLANNER_COLLATIONS_MAP = Collections.synchronizedMap(new WeakHashMap<>());
+
+  /**
+   * Sets the current planner for the optimization pass.
+   * Should be called at the beginning of optimization to enable
+   * proper trait satisfaction checking with original collations.
+   */
+  public static void setPlannerContext(RelOptPlanner planner) {
+    CURRENT_PLANNER.set(planner);
+  }
+
+  /**
+   * Clears the current planner context.
+   * Should be called at the end of optimization.
+   */
+  public static void clearPlannerContext() {
```

Comment:
Q: this kind of usage reminds me of the `AutoCloseable` interface + `try-with-resources`, but making this class implement that would be too invasive probably?